name-as/src/nma.rs

```diff
@@ -85,31 +87,91 @@ pub fn r_operation(mnemonic: &str) -> Result<R, &'static str> {
 /// Parses an I-type instruction mnemonic into an [I]
 pub fn i_operation(mnemonic: &str) -> Result<I, &'static str> {
     match mnemonic {
+        "ori" => Ok(I {
+            opcode: 0xd,
+            form: IForm::RtRsImm,
+        }),
+        "lb" => Ok(I {
+            opcode: 0x20,
+            form: IForm::RtImmRs,
+        }),
+        "lbu" => Ok(I {
+            opcode: 0x24,
+            form: IForm::RtImmRs,
+        }),
+        "lh" => Ok(I {
+            opcode: 0x21,
+            form: IForm::RtImmRs,
+        }),
+        "lhu" => Ok(I {
+            opcode: 0x25,
+            form: IForm::RtImmRs,
+        }),
+        "lw" => Ok(I {
+            opcode: 0x23,
+            form: IForm::RtImmRs,
+        }),
+        "ll" => Ok(I {
+            opcode: 0x30,
+            form: IForm::RtImmRs,
+        }),
         "lui" => Ok(I {
             opcode: 0xf,
             form: IForm::RtImm,
         }),
+        "sb" => Ok(I {
+            opcode: 0x28,
+            form: IForm::RtImmRs,
+        }),
+        "sh" => Ok(I {
+            opcode: 0x29,
+            form: IForm::RtImmRs,
+        }),
+        "sw" => Ok(I {
+            opcode: 0x2b,
+            form: IForm::RtImmRs,
+        }),
+        "sc" => Ok(I {
+            opcode: 0x38,
+            form: IForm::RtImmRs,
+        }),
+        "beq" => Ok(I {
+            opcode: 0x4,
+            form: IForm::RsRtLabel,
+        }),
+        "bne" => Ok(I {
+            opcode: 0x5,
+            form: IForm::RsRtLabel,
+        }),
         _ => Err("Failed to match I-instr mnemonic"),
     }
 }
 
-/// Split a string into meaningful, atomic elements of the MIPS language
-pub fn tokenize(raw_text: &str) -> Vec<&str> {
-    // raw_text.split_whitespace().collect::<Vec<&str>>()
-    raw_text.split(&[',', ' ', '\t', '\r', '\n'][..])
-        .filter(|&s| !s.is_empty())
-        .collect::<Vec<&str>>()
+/// Parses a J-type instruction mnemonic into a [J]
+fn j_operation(mnemonic: &str) -> Result<J, &'static str> {
+    match mnemonic {
+        "j" => Ok(J { opcode: 0x2 }),
+        "jal" => Ok(J { opcode: 0x3 }),
+        _ => Err("Failed to match J-instr mnemonic"),
+    }
 }
 
 /// Write a u32 into a file, zero-padded to 32 bits (4 bytes)
 pub fn write_u32(mut file: &File, data: u32) -> std::io::Result<()> {
+    fn convert_endianness(input: u32) -> u32 {
+        ((input & 0x000000FF) << 24)
+            | ((input & 0x0000FF00) << 8)
+            | ((input & 0x00FF0000) >> 8)
+            | ((input & 0xFF000000) >> 24)
+    }
+
     const PADDED_LENGTH: usize = 4;
 
     // Create a 4-length buffer
     let mut padded_buffer: [u8; PADDED_LENGTH] = [0; PADDED_LENGTH];
 
     // Convert data into bytes
-    let bytes: [u8; PADDED_LENGTH] = data.to_be_bytes();
+    let bytes: [u8; PADDED_LENGTH] = (convert_endianness(data)).to_be_bytes();
```

Comment:
I dumped the assembled code from MARS and it came out as little-endian, so I started dumping as little-endian so that I could hash my output to compare it to MARS. Should I make that a test feature only?

Reply:
Resolved